### PR TITLE
Fix  access to `_msvccompiler` from newer distutils

### LINKED ISSUE #141319: AttributeError: module 'distutils' has no attribute '_msvccompiler'
### 🐛 Describe the bug

When we trigger Windows CI, we happen to encounter a build failure and error message is as follows:
```bash
AttributeError: module 'distutils' has no attribute '_msvccompiler'
```

Since the `_msvccompiler` has bee deprecated in `setuptools` from `72.1.0`(exclusive) and Intel team submitted a PR to fix the version of `setuptools` - https://github.com/pytorch/pytorch/pull/136489

It seems like Windows CI scripts do not apply this fix by sticking `setuptools` to a particular compatible version.

### Versions

Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Enterprise (10.0.22631 64-bit)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.11 (tags/v3.10.11:7d4cc5a, Apr  5 2023, 00:38:17) [MSC v.1929 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-10-10.0.22631-SP0
Is CUDA available: N/A
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

CPU:
Name: 12th Gen Intel(R) Core(TM) i7-1270P
Manufacturer: GenuineIntel
Family: 198
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 1711
MaxClockSpeed: 2200
L2CacheSize: 9216
L2CacheSpeed: None
Revision: None

Versions of relevant libraries:
[pip3] torch==2.5.0a0+git32f585d
[conda] Could not collect

cc @peterjc123 @mszhanyi @skyline75489 @nbcsm @iremyux @Blackhex @seemethere @malfet @pytorch/pytorch-dev-infra

### PR DESCRIPTION
**Pull request by malfet**
Newer versions of distutils no longer import `_msvccompiler` upon init(on Windows platform, that was not the case on other platforms even before 74), but it's still accessible if one chooses to import it directly.
Test plan:
```
% python -c 'from setuptools import distutils; print(distutils.__version__, hasattr(distutils, "_msvccompiler")); from distutils import _msvccompiler; import setuptools; print(setuptools.__version__, _msvccompiler.__file__)' 
3.10.9 False
65.5.0 /usr/local/fbcode/platform010/Python3.10.framework/Versions/3.10/lib/python3.10/site-packages/setuptools/_distutils/_msvccompiler.py
```
and
```
% python -c 'from setuptools import distutils; print(distutils.__version__, hasattr(distutils, "_msvccompiler")); from distutils import _msvccompiler; import setuptools; print(setuptools.__version__, _msvccompiler.__file__)' 
3.13.0 False
75.6.0 /Users/malfet/py312-venv/lib/python3.13/site-packages/setuptools/_distutils/_msvccompiler.py
```

Gave up trying to appease the linker, so rewrote it as following function:
```python
def _get_vc_env(vc_arch: str) -> dict[str, str]:
    try:
        from setuptools import distutils  # type: ignore[import]

        return distutils._msvccompiler._get_vc_env(vc_arch)  # type: ignore[no-any-return]
    except AttributeError:
        from setuptools._distutils import _msvccompiler  #type: ignore[import]

        return _msvccompiler._get_vc_env(vc_arch)  # type: ignore[no-any-return]
```

This PR also undoes setuptools version restriction introduced by  https://github.com/pytorch/pytorch/pull/136489 as premise for restriction is incorrect

Fixes https://github.com/pytorch/pytorch/issues/141319

### CODE REVIEW

**Review by huydhn — APPROVED** [review]
LGTM! I wonder if this this will go away eventually that explicit import won't work anymore (could be the reason why it's hidden now)

**Review by atalman — APPROVED** [review]
lgtm

File: tools/build_pytorch_libs.py [hunk 1]
```diff
@@ -4,13 +4,14 @@
...
 
-from setuptools import distutils  # type: ignore[import]
 
 from .setup_helpers.cmake import CMake, USE_NINJA
 from .setup_helpers.env import check_negative_env_flag, IS_64BIT, IS_WINDOWS
 
 
 def _overlay_windows_vcvars(env: dict[str, str]) -> dict[str, str]:
+    from setuptools import distutils  # type: ignore[import]
+    from distutils import _msvccompiler  # type: ignore[import]
```
Comment (malfet) [inline tools/build_pytorch_libs.py:14]:
```suggestion
    from distutils import _msvccompiler  # type: ignore[attr-defined]
```

**Comment by EikanWang** [comment]
@malfet , we just landed the PR(https://github.com/pytorch/pytorch/pull/141086) to fix https://github.com/pytorch/pytorch/issues/141268 for Intel GPU. I intend to pass the CI failure of this PR through rebasing.

**Pull request closed without merging**